### ISSUE TITLE
[GSoC] [Discussion] Implementation of composition and inversion of formal power series

### ISSUE THREAD

**Issue by ArighnaIITG** [sympy/sympy#16975]
This issue is related to the implementation of `composition and inversion of formal power series`.

For `composition`, there exists a closed form expression for the resultant power series, where the resultant coefficient sequence `ak` involves Bell polynomials, as follows --

![image](https://user-images.githubusercontent.com/28482640/58986902-ffd1ee80-87fb-11e9-95d5-bf4965dd4c1c.png)

But, this will be quite hard to implement. 
I was wondering, if we can write a function, which will not return a `Formal Power Series` object, but only the first `n+1` coefficients for the composed series. Fast algorithms do exist for that, as [mentioned here](http://www.eecs.harvard.edu/~htk/publication/1978-jacm-brent-kung.pdf).

Or, if we need to implement the Bell Polynomial closed form expression, how can we proceed? 

Same thing exists for `inversion` also. For `inversion`, there is no `fast algorithm` as such. A recurrence relation exists for calculating the closed form expression of `ak`(reference [here](https://math.stackexchange.com/questions/710252/multiplicative-inverse-of-a-power-series)). I know of no such algorithm that will calculate the first `n+1` coefficients of a inverted formal power series.

This issue is to discuss, if we can just calculate the n+1 coefficients, without the need to make a separate Formal Power Series object.

**Comment by ArighnaIITG** [comment]
ping @leosartaj @asmeurer @jksuom @smichr

**Comment by leosartaj** [comment]
> This issue is related to the implementation of `composition and inversion of formal power series`.
> 
> For `composition`, there exists a closed form expression for the resultant power series, where the resultant coefficient sequence `ak` involves Bell polynomials, as follows --

I don't think this is closed form as we expect for current ``FormalPowerSeries`` implementation. This formula is dependent on older terms, and each coefficient cannot be computed independently.
> 
> ![image](https://user-images.githubusercontent.com/28482640/58986902-ffd1ee80-87fb-11e9-95d5-bf4965dd4c1c.png)
> 
> But, this will be quite hard to implement.
> I was wondering, if we can write a function, which will not return a `Formal Power Series` object, but only the first `n+1` coefficients for the composed series. Fast algorithms do exist for that, as [mentioned here](http://www.eecs.harvard.edu/~htk/publication/1978-jacm-brent-kung.pdf).

I think it can be a start. Something like.

```
>>> f = fps(sin(x), x)
>>> g = fps(cos(x), x)
>>> f.compose(g, order=4) # need to give order
.... + O(x**4)
```

> 
> Or, if we need to implement the Bell Polynomial closed form expression, how can we proceed?
> 

It can be a good thing to let ``compose`` return a maybe ``FormalPowerSeriesFinite`` or ``FormalPowerSeriesCompose`` object derived from ``FormalPowerSeries`` class. I think some of the methods can work as it is. Computation of coefficients would be slower since it will no longer be a formula not dependent on previous terms. How about making a Bell Polynomials sequence that is cached? That should speed things up. 

> Same thing exists for `inversion` also. For `inversion`, there is no `fast algorithm` as such. A recurrence relation exists for calculating the closed form expression of `ak`(reference [here](https://math.stackexchange.com/questions/710252/multiplicative-inverse-of-a-power-series)). I know of no such algorithm that will calculate the first `n+1` coefficients of a inverted formal power series.
> 
> This issue is to discuss, if we can just calculate the n+1 coefficients, without the need to make a separate Formal Power Series object.

This can be tackled on similar lines like ``composition`` . I feel it's very close to ``composition`` problem.

**Comment by ArighnaIITG** [comment]
> Computation of coefficients would be slower since it will no longer be a formula not dependent on previous terms. How about making a Bell Polynomials sequence that is cached? That should speed things up.

I couldn't get you on this. Can you kindly elaborate? Even if we create a new class for it, we still do think about the coefficient sequence, and the powers of x, right?

**Comment by leosartaj** [comment]
> > Computation of coefficients would be slower since it will no longer be a formula not dependent on previous terms. How about making a Bell Polynomials sequence that is cached? That should speed things up.
> 
> I couldn't get you on this. Can you kindly elaborate? Even if we create a new class for it, we still do think about the coefficient sequence, and the powers of x, right?

Okay, Let's try it this way,

So we need to represent g(f(x))

Thinking it as ``FormalPowerSeries`` object we can think it having some ``tk`` and ``xk``.  ``xk`` can be just ``x**k`` or ``x**k / k!`` (dependent on what is more perform-ant vs ease of change). Let's say ``xk`` is ``x**k/k!`` just for simplicity.

Then ``tk`` comes out to be ``Sum(bk * B_n_k(a1...a(n-k+1), (k, 1, n))`` # dependent on n.

Computing ``tk`` is what needs to be worked out. If you see ``tk`` is dependent on a series of ``ak``. Computing this is the trickiest part.

I would suggest to get a sequence of bell polynomials (sequences are cached so should be cached) i.e just work with term ``B_n_k(a1...a(n-k+1)``. After that it's just a finite sum of multiplying two sequences ``bk`` and ``B_n_k(a1...a(n-k+1)``. (how to represent such kind of sequence?)

This is just my idea on it.  This will need tweaking depending on actual code.

Let me know your thoughts on this. It will be great if you can come up with an api design along the lines of your proposal or my idea.

**Comment by leosartaj** [comment]
Implemented in #17064. Closing.
